include/merlin/array_kernels.cuh

```diff
@@ -0,0 +1,120 @@
+  if (tid < n) {
+    data[tid] = val;
+  }
+}
+
+template <typename K>
+__global__ void keys_not_empty(const K* keys, bool* masks, size_t n) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  if (tid < n) {
+    masks[tid] = keys[tid] == EMPTY_KEY ? true : false;
+  }
+}
+
+template <typename Tidx, int TILE_SIZE = 8>
+__global__ void gpu_cell_count(const bool* masks, Tidx* offsets, size_t n,
+                               size_t* n_existed) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  auto g = cg::tiled_partition<TILE_SIZE>(cg::this_thread_block());
+  int rank = g.thread_rank();
+  bool is_existed = false;
+  if (tid < n) {
+    if (masks[tid]) {
+      is_existed = true;
+    }
+  }
+  unsigned int vote = g.ballot(is_existed);
+  int g_ones = __popc((int)vote);
+  if (rank == 0 && tid < n) {
+    offsets[tid / TILE_SIZE] = static_cast<Tidx>(g_ones);
+    atomicAdd(static_cast<uint64_t*>(n_existed), static_cast<uint64_t>(g_ones));
+  }
+}
+
+template <typename K, typename V, typename M, typename Tidx, int TILE_SIZE = 8>
+__global__ void gpu_select_kvm_kernel(const bool* masks, size_t n,
+                                      const Tidx* offsets, K* __restrict keys,
+                                      V* __restrict values, M* __restrict metas,
+                                      const size_t dim) {
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+  auto g = cg::tiled_partition<TILE_SIZE>(cg::this_thread_block());
+  int rank = g.thread_rank();
+
+  bool is_existed = false;
+  if (tid < n) {
+    if (masks[tid]) {
+      is_existed = true;
+    }
+  }
+  unsigned int vote = g.ballot(is_existed);
+  unsigned int r_vote = __brev(vote) >> (32 - TILE_SIZE);
+
+  if (tid < n) {
+    r_vote = r_vote >> (TILE_SIZE - rank - 1);
+    if (masks[tid]) {
+      int prefix_n = __popc(r_vote) - 1;
+      Tidx bias = offsets[tid / TILE_SIZE] + static_cast<Tidx>(prefix_n);
+      keys[bias] = keys[tid];
+      metas[bias] = metas[tid];
+      for (size_t j = 0; j < dim; j++) {
+        values[dim * bias + j] = keys[dim * tid + j];
+      }
+    }
+  }
+}
+
+template <typename K, typename V, typename M, typename Tidx, int TILE_SIZE = 8>
+void gpu_boolean_mask(size_t grid_size, size_t block_size, const bool* masks,
+                      size_t n, size_t* n_evicted, Tidx* offsets,
+                      K* __restrict keys, V* __restrict values,
+                      M* __restrict metas, size_t dim, cudaStream_t stream) {
+  size_t n_offsets = (n + TILE_SIZE - 1) / TILE_SIZE;
+  gpu_cell_count<Tidx, TILE_SIZE>
+      <<<grid_size, block_size, 0, stream>>>(masks, offsets, n, n_evicted);
+#if THRUST_VERSION >= 101600
+  auto policy = thrust::cuda::par_nosync.on(stream);
+#else
+  auto policy = thrust::cuda::par.on(stream);
+#endif
+  thrust::device_ptr<Tidx> d_src(offsets);
+  thrust::device_ptr<Tidx> d_dest(offsets);
+  thrust::exclusive_scan(policy, d_src, d_src + n_offsets, d_dest);
+  gpu_select_kvm_kernel<K, V, M, Tidx, TILE_SIZE>
```

Comment:
Maybe using the MERLIN_CHECK to confirm the `TILE_SIZE <=32` before calling `gpu_select_kvm_kernel` would be better.